node/overseer/overseer-gen/src/lib.rs

```diff
@@ -196,13 +196,14 @@ pub struct SignalsReceived(Arc<AtomicUsize>);
 impl SignalsReceived {
 	/// Load the current value of received signals.
 	pub fn load(&self) -> usize {
-		// off by a few is ok
-		self.0.load(atomic::Ordering::Relaxed)
+		// It's imperative that we don't underestimate the amount
+		// of signals received.
+		self.0.load(atomic::Ordering::Acquire)
 	}
 
 	/// Increase the number of signals by one.
 	pub fn inc(&self) {
-		self.0.fetch_add(1, atomic::Ordering::Acquire);
+		self.0.fetch_add(1, atomic::Ordering::AcqRel);
```

Comment:
Could you go into more detail why?

Reply:
> If an atomic store in thread A is tagged memory_order_release and an atomic load in thread B from the same variable is tagged memory_order_acquire, all memory writes (non-atomic and relaxed atomic) that happened-before the atomic store from the point of view of thread A, become visible side-effects in thread B.

I suppose it's that the atomic write itself isn't counted?

Reply:
No, SeqCst is only necessary if you have multiple threads whose execution must be observed in the same order.

We only have one variable, so acquire/release is sufficient. I'll expand on this in a bit

Reply:
AFAIR, there will be no difference on x86_64 - all RMW (and increment is an example of such an operation) atomic operations are promoted to `SeqCst` by `lock` prefix. I'm not sure about the compiler assumptions on that but I believe there will be no difference on that side as well. In general, the strict `SeqCst` order is really needed in a very few use cases but it is the most safe option in case of doubts. Since there is no difference on x86_64 which is our main target platform, I would also suggest to use `SeqCst`

Reply:
Well, even `Relaxed` is promoted to `lock` for RMW, so there is no difference on x86: https://godbolt.org/z/4hzPrqKz1

Reply:
> No, SeqCst is only necessary if you have multiple threads whose execution must be observed in the same order.
> 
> We only have one variable, so acquire/release is sufficient. I'll expand on this in a bit

Using sequential consistency comes at a cost, but all threads observe the same deterministic ordering of memory reads/writes to the memory location. It would not be helpful as you said in this context where there is a single producer/consumer theads, but,I was unsure of the contexts from which inc() is called based on the PR description, so I errored on the safe side. 

Also as @vstakhov pointed out, using SeqCst would make the code work as intended on other platforms with weaker memory ordering like `aarch64`:

<img width="596" alt="Screen-Shot-2021-03-26-at-10 21 13-AM" src="https://user-images.githubusercontent.com/54316454/158559536-ba937bce-909c-418f-b67b-0dd69a0ccb7d.png">

Reply:
There is no point in using `SeqCst` with just one atomic. All stores with `Release` are seen by loads of the same var with `Acquire` - I just re-read the rust for rustaceans (print) and https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html to assert on that.

Reply:
> Well, even `Relaxed` is promoted to `lock` for RMW, so there is no difference on x86: https://godbolt.org/z/4hzPrqKz1

Could it be that the whole atomics are optimized away - the program shown is just a single thread after all.

Reply:
> AFAIR, there will be no difference on x86_64 - all RMW (and increment is an example of such an operation) atomic operations are promoted to `SeqCst` by `lock` prefix. I'm not sure about the compiler assumptions on that but I believe there will be no difference on that side as well. In general, the strict `SeqCst` order is really needed in a very few use cases but it is the most safe option in case of doubts. Since there is no difference on x86_64 which is our main target platform, I would also suggest to use `SeqCst`

While you're absolutely correct that x86_64 has a few hardware features that makes i.e. Acq/Release semantics as cheap as relaxed, we don't want to special case for x86_64

Reply:
Note that thread scheduling is not the same as re-ordering. Since we only have one variable, there are no atomics that could be re-ordered and hence cause trouble. All stores to `SignalsReceived.0` are visible to loads by the `Acquire`/`Release` semantics.